File: ios/Algo/Features/Settings/ViewModels/SettingsViewModel.swift
// SettingsViewModel.swift
// Algo
//
// ViewModel for the Settings screen. Manages user info,
// notification preferences, and logout functionality.

import Foundation

// MARK: - UserInfo

/// Represents the currently authenticated user's profile information.
struct UserInfo: Codable, Sendable {
    let id: String
    let email: String
    let name: String
    let role: String
    let tenantId: String
    let avatarUrl: String?

    enum CodingKeys: String, CodingKey {
        case id, email, name, role
        case tenantId = "tenant_id"
        case avatarUrl = "avatar_url"
    }
}

// MARK: - NotificationPreference

/// Notification settings for a single alert type.
struct NotificationPreference: Codable, Sendable, Identifiable {
    var id: String { alertType }

    /// The alert type key (e.g., "hot_lead", "campaign_drop").
    let alertType: String

    /// Whether this alert type is enabled.
    var enabled: Bool

    /// Whether push notifications are sent for this type.
    var push: Bool

    /// Whether email notifications are sent for this type.
    var email: Bool

    /// Optional threshold percentage (used by budget_alert).
    var thresholdPct: Int?

    /// Optional absolute threshold (used by credit_low).
    var threshold: Int?
}

// MARK: - NotificationSettings

/// Complete notification settings for all alert types.
struct NotificationSettings: Codable, Sendable {
    var hotLead: NotificationPreference
    var campaignDrop: NotificationPreference
    var budgetAlert: NotificationPreference
    var revenueSpike: NotificationPreference
    var creditLow: NotificationPreference
    var followupOverdue: NotificationPreference

    /// Returns all preferences as an ordered array for iteration.
    var allPreferences: [NotificationPreference] {
        [hotLead, campaignDrop, budgetAlert, revenueSpike, creditLow, followupOverdue]
    }

    /// Default notification settings with all alerts enabled.
    static let `default` = NotificationSettings(
        hotLead: NotificationPreference(
            alertType: "hot_lead", enabled: true, push: true, email: true
        ),
        campaignDrop: NotificationPreference(
            alertType: "campaign_drop", enabled: true, push: true, email: false
        ),
        budgetAlert: NotificationPreference(
            alertType: "budget_alert", enabled: true, push: true, email: true,
            thresholdPct: 80
        ),
        revenueSpike: NotificationPreference(
            alertType: "revenue_spike", enabled: true, push: false, email: true
        ),
        creditLow: NotificationPreference(
            alertType: "credit_low", enabled: true, push: true, email: true,
            threshold: 500
        ),
        followupOverdue: NotificationPreference(
            alertType: "followup_overdue", enabled: true, push: true, email: false
        )
    )
}

// MARK: - SettingsViewModel

/// Manages the state and business logic for the Settings screen.
///
/// Responsibilities:
/// - Loads and displays current user info
/// - Manages notification preference settings
/// - Handles logout flow
@MainActor
final class SettingsViewModel: ObservableObject {

    // MARK: Published State

    /// Current user profile information.
    @Published private(set) var user: UserInfo?

    /// Notification settings for all alert types.
    @Published var notificationSettings: NotificationSettings = .default

    /// Whether data is being loaded.
    @Published private(set) var isLoading = false

    /// Whether settings are being saved.
    @Published private(set) var isSaving = false

    /// Current error message, if any.
    @Published private(set) var error: String?

    /// Whether the user has been successfully logged out.
    @Published private(set) var didLogout = false

    // MARK: Dependencies

    private let baseURL: URL
    private let session: URLSession
    private let tokenProvider: () -> String?
    private let onLogout: () -> Void

    // MARK: Initialization

    /// Creates a new `SettingsViewModel`.
    /// - Parameters:
    ///   - baseURL: The base URL for the Algo API.
    ///   - session: URL session for requests.
    ///   - tokenProvider: Closure returning the current JWT token.
    ///   - onLogout: Closure called when the user logs out.
    init(
        baseURL: URL = URL(string: "https://api.algo.algonit.com/v1")!,
        session: URLSession = .shared,
        tokenProvider: @escaping @Sendable () -> String?,
        onLogout: @escaping () -> Void
    ) {
        self.baseURL = baseURL
        self.session = session
        self.tokenProvider = tokenProvider
        self.onLogout = onLogout
    }

    // MARK: Public Methods

    /// Loads user info and notification settings from the backend.
    func loadSettings() async {
        isLoading = true
        error = nil

        // In a production app, these would be separate API calls.
        // For now, user info is typically cached from the login response.
        // Notification settings would come from GET /notifications/settings.

        isLoading = false
    }

    /// Sets the user info (typically called after login).
    /// - Parameter user: The authenticated user's profile.
    func setUser(_ user: UserInfo) {
        self.user = user
    }

    /// Updates notification settings on the backend.
    func updateNotificationSettings() async {
        isSaving = true
        error = nil

        do {
            let url = baseURL.appendingPathComponent("notifications/settings")
            var request = URLRequest(url: url)
            request.httpMethod = "PUT"
            request.setValue("application/json", forHTTPHeaderField: "Content-Type")
            request.setValue(UUID().uuidString, forHTTPHeaderField: "X-Request-ID")

            if let token = tokenProvider() {
                request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
            }

            let payload = buildNotificationSettingsPayload()
            request.httpBody = try JSONSerialization.data(withJSONObject: payload)

            let (_, response) = try await session.data(for: request)

            guard let httpResponse = response as? HTTPURLResponse,
                  (200...299).contains(httpResponse.statusCode) else {
                throw SettingsError.saveFailed
            }
        } catch {
            self.error = "Failed to save notification settings."
        }

        isSaving = false
    }

    /// Logs the user out by calling the backend and clearing local state.
    func logout() async {
        do {
            let url = baseURL.appendingPathComponent("auth/logout")
            var request = URLRequest(url: url)
            request.httpMethod = "POST"
            request.setValue("application/json", forHTTPHeaderField: "Content-Type")

            if let token = tokenProvider() {
                request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
            }

            let deviceId = UIDevice.current.identifierForVendor?.uuidString ?? ""
            let body = ["device_id": deviceId]
            request.httpBody = try JSONEncoder().encode(body)

            // Fire and forget - we log out locally regardless
            _ = try? await session.data(for: request)
        } catch {
            // Proceed with local logout even if the network call fails
        }

        didLogout = true
        onLogout()
    }

    // MARK: Private Helpers

    /// Builds the notification settings payload for the API.
    private func buildNotificationSettingsPayload() -> [String: Any] {
        var payload: [String: Any] = [:]

        func prefDict(_ pref: NotificationPreference) -> [String: Any] {
            var dict: [String: Any] = [
                "enabled": pref.enabled,
                "push": pref.push,
                "email": pref.email
            ]
            if let thresholdPct = pref.thresholdPct {
                dict["threshold_pct"] = thresholdPct
            }
            if let threshold = pref.threshold {
                dict["threshold"] = threshold
            }
            return dict
        }

        payload["hot_lead"] = prefDict(notificationSettings.hotLead)
        payload["campaign_drop"] = prefDict(notificationSettings.campaignDrop)
        payload["budget_alert"] = prefDict(notificationSettings.budgetAlert)
        payload["revenue_spike"] = prefDict(notificationSettings.revenueSpike)
        payload["credit_low"] = prefDict(notificationSettings.creditLow)
        payload["followup_overdue"] = prefDict(notificationSettings.followupOverdue)

        return payload
    }
}

// MARK: - SettingsError

enum SettingsError: LocalizedError {
    case saveFailed

    var errorDescription: String? {
        switch self {
        case .saveFailed:
            return "Failed to save settings. Please try again."
        }
    }
}

// MARK: - UIDevice Import

import UIKit

File: android/app/src/main/java/com/algonit/algo/features/settings/presentation/SettingsViewModel.kt
package com.algonit.algo.features.settings.presentation

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.algonit.algo.features.auth.domain.repository.AuthRepository
import com.algonit.algo.ui.theme.DarkModeManager
import com.algonit.algo.ui.theme.ThemeMode
import dagger.hilt.android.lifecycle.HiltViewModel
import io.ktor.client.HttpClient
import io.ktor.client.call.body
import io.ktor.client.request.get
import io.ktor.client.request.put
import io.ktor.client.request.setBody
import io.ktor.http.ContentType
import io.ktor.http.contentType
import javax.inject.Named
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import javax.inject.Inject

@Serializable
data class UserInfo(
    val id: String,
    val email: String,
    val name: String,
    val role: String,
    @SerialName("tenant_id") val tenantId: String,
    @SerialName("avatar_url") val avatarUrl: String? = null
)

@Serializable
data class AlertPreference(
    val enabled: Boolean = true,
    val push: Boolean = true,
    val email: Boolean = false,
    @SerialName("threshold_pct") val thresholdPct: Int? = null,
    val threshold: Int? = null
)

@Serializable
data class NotificationSettings(
    @SerialName("hot_lead") val hotLead: AlertPreference = AlertPreference(),
    @SerialName("campaign_drop") val campaignDrop: AlertPreference = AlertPreference(),
    @SerialName("budget_alert") val budgetAlert: AlertPreference = AlertPreference(
        thresholdPct = 80
    ),
    @SerialName("revenue_spike") val revenueSpike: AlertPreference = AlertPreference(),
    @SerialName("credit_low") val creditLow: AlertPreference = AlertPreference(
        threshold = 500
    ),
    @SerialName("followup_overdue") val followupOverdue: AlertPreference = AlertPreference()
)

data class SettingsUiState(
    val user: UserInfo? = null,
    val notificationSettings: NotificationSettings = NotificationSettings(),
    val themeMode: ThemeMode = ThemeMode.SYSTEM,
    val isLoading: Boolean = false,
    val isSaving: Boolean = false,
    val error: String? = null,
    val saveSuccess: Boolean = false
)

@HiltViewModel
class SettingsViewModel @Inject constructor(
    @Named("api") private val httpClient: HttpClient,
    private val darkModeManager: DarkModeManager,
    private val authRepository: AuthRepository
) : ViewModel() {

    private val _uiState = MutableStateFlow(SettingsUiState())
    val uiState: StateFlow<SettingsUiState> = _uiState.asStateFlow()

    init {
        loadSettings()
        observeThemeMode()
    }

    /**
     * Observes the persisted theme mode and updates the UI state reactively.
     */
    private fun observeThemeMode() {
        viewModelScope.launch {
            darkModeManager.themeMode.collect { mode ->
                _uiState.update { it.copy(themeMode = mode) }
            }
        }
    }

    /**
     * Sets the app theme mode and persists the preference.
     */
    fun setThemeMode(mode: ThemeMode) {
        viewModelScope.launch {
            darkModeManager.setThemeMode(mode)
        }
    }

    fun loadSettings() {
        viewModelScope.launch {
            _uiState.update { it.copy(isLoading = true, error = null) }

            try {
                // Load user info from stored token/profile
                // In production this would come from SecureStorage or a profile endpoint
                val notificationSettings = try {
                    httpClient.get("/notifications/settings").body<NotificationSettings>()
                } catch (_: Exception) {
                    NotificationSettings() // Use defaults if endpoint fails
                }

                _uiState.update {
                    it.copy(
                        notificationSettings = notificationSettings,
                        isLoading = false,
                        error = null
                    )
                }
            } catch (e: Exception) {
                _uiState.update {
                    it.copy(
                        isLoading = false,
                        error = e.message ?: "Failed to load settings"
                    )
                }
            }
        }
    }

    fun updateNotificationSettings(settings: NotificationSettings) {
        viewModelScope.launch {
            _uiState.update { it.copy(isSaving = true, saveSuccess = false) }

            try {
                httpClient.put("/notifications/settings") {
                    contentType(ContentType.Application.Json)
                    setBody(settings)
                }

                _uiState.update {
                    it.copy(
                        notificationSettings = settings,
                        isSaving = false,
                        saveSuccess = true,
                        error = null
                    )
                }
            } catch (e: Exception) {
                _uiState.update {
                    it.copy(
                        isSaving = false,
                        error = e.message ?: "Failed to save settings"
                    )
                }
            }
        }
    }

    fun setUser(user: UserInfo) {
        _uiState.update { it.copy(user = user) }
    }

    fun dismissSaveSuccess() {
        _uiState.update { it.copy(saveSuccess = false) }
    }

    fun dismissError() {
        _uiState.update { it.copy(error = null) }
    }

    fun logout() {
        viewModelScope.launch {
            // AuthRepository.logout() handles both the API call and clearing secure storage
            authRepository.logout()

            // Clear local settings state
            _uiState.update {
                SettingsUiState()
            }
        }
    }
}
